### PR TITLE
Add `adjustOpacity` action

### LINKED ISSUE #11205: Action for adjusting opacity at runtime
# Description of the new feature/enhancement
I love the pane of glass look for Terminal. However, when using a dark theme for example in Terminal and the window is in front-of a light colored windows (i.e. Browser with a white page), the text is unreadable  in terminal. And when the Terminal is in front of a dark background the text color in Terminal is nice and crisp. What would be great is if the Terminal can detect the underlying window color and dynamically adjust the opacity to ensure the text of the Terminal window remains readable regardless of the window that Terminal is in front of.

Can't read text in Terminal with transparency in front of a light window when using a dark theme. In order to keep the pane of glass look, I need to keep adjusting the transparency setting based on the background colors (background theme and/or window the Terminal is overlaying). New functionality would automatically adjust this or at a minimum provide a hot key to adjust transparency on demand

# Proposed technical implementation details (optional)

### PR DESCRIPTION
**Pull request by zadjii-msft**
Adds an action which can be used to change the opacity at runtime. This is a follow up to some of the other work I had been doing around opacity and settings previewing at the end of the year.

> edit: pseudo-spec
> 
> * `adjustOpacity`:
>   * `"opacity"`: (**required**) an integer number to set the opacity to.
>   * `"relative"`: (defaults to `true`)
>     * If false, set the opacity to the given value, which should be between [0, 100].
>     * If true, then use `opacity` as a relative adjustment to the current opacity. So the implementation would get the current opacity for this pane, then add this action's opacity to that value. Should be between [-100, 100]
> 
> This would allow both setting exactly 25% opacity with an action, and binding an action that increases/decreases the opacity one {amount}

It's preview-able too, which is neat. 

adjust-opacity-preview-001

* [x] Closes #11205
* [x] I work here
* [x] Docs updated: https://github.com/MicrosoftDocs/terminal/pull/477

### CODE REVIEW

File: src/renderer/dx/DxRenderer.cpp [hunk 1]
```diff
@@ -910,6 +910,15 @@ void DxEngine::_ReleaseDeviceResources() noexcept
...
     // grayscale antialiasing)
+    //
+    // October 2021: We're no longer forcing the BG of the run to be opaque when
+    // cleartype is enabled and the background isn't fully opaque. In the case
+    // of (!useAcrylic && opacity<1.0), we actually can still render cleartype
+    // text just fine. It's only acrylic that messes us up. So we're making a
+    // small change. Now, when the user requests acrylic, text that's rendered
+    // on the default text BG will always use grayscale, rather than cleartype.
+    // This helps support the scenario where the user has (useAcrylic &&
+    // opacity==1.0)
```
Comment (carlos-zamora) [inline src/renderer/dx/DxRenderer.cpp:921]: What's the deal with this comment? Did we accidentally delete it earlier?

File: src/cascadia/TerminalControl/ControlSettings.h [hunk 2]
```diff
@@ -65,6 +65,7 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         // appearance is being used should be more careful. Fortunately, this
         // situation is generally only used when a control is first created, or
         // when calling UpdateSettings.
+
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlSettings.h:68]: looks like you can remove this

File: src/cascadia/TerminalControl/ControlCore.h [hunk 3]
```diff
@@ -23,6 +23,8 @@
 
 #include <til/ticket_lock.h>
 
+#include <til/ticket_lock.h>
+
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.h:27]: duplicate with include statement above

File: src/cascadia/TerminalApp/ActionPreviewHandlers.cpp [hunk 4]
```diff
@@ -65,7 +66,7 @@ namespace winrt::TerminalApp::implementation
     // - <none>
     // Return Value:
     // - <none>
-    void TerminalPage::_EndPreviewColorScheme()
+    void TerminalPage::_RunRestorePreviews()
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/ActionPreviewHandlers.cpp:69]: looks like you need to update the function comment

File: src/cascadia/TerminalApp/ActionPreviewHandlers.cpp [hunk 5]
```diff
@@ -111,6 +112,27 @@ namespace winrt::TerminalApp::implementation
         }
     }
 
+    void TerminalPage::_PreviewAdjustOpacity(const Settings::Model::AdjustOpacityArgs& args)
+    {
+        // Clear the saved preview funcs because we don't need to add a restore each time
+        // the preview color changes, we only need to be able to restore the last one.
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/ActionPreviewHandlers.cpp:118]:
```suggestion
        // the preview changes, we only need to be able to restore the last one.
```

File: src/cascadia/TerminalApp/ActionPreviewHandlers.cpp [hunk 6]
```diff
@@ -111,6 +112,27 @@ namespace winrt::TerminalApp::implementation
...
     }
 
+    void TerminalPage::_PreviewAdjustOpacity(const Settings::Model::AdjustOpacityArgs& args)
+    {
+        // Clear the saved preview funcs because we don't need to add a restore each time
+        // the preview color changes, we only need to be able to restore the last one.
+        _restorePreviewFuncs.clear();
+
+        _ApplyToActiveControls([&](const auto& control) {
+            // Stash a copy of the original scheme.
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/ActionPreviewHandlers.cpp:122]:
```suggestion
            // Stash a copy of the original
```

File: doc/cascadia/profiles.schema.json [hunk 7]
```diff
@@ -1264,6 +1265,34 @@
...
+              "const": "adjustOpacity"
+            },
+            "opacity": {
+              "type": "integer",
+              "minimum": -100,
+              "maximum": 100,
+              "default": 0,
+              "description": "How opaque the terminal should become or how much the opacity should be changed by, depending on the value of `relative`"
+            },
+            "opacity": {
```
Comment (carlos-zamora) [inline doc/cascadia/profiles.schema.json:1287]:
```suggestion
            "relative": {
```

**Review by carlos-zamora — COMMENTED** [review]
Love the design. Half-related question: can we update `adjustFontSize` to operate the same way?

I'm mildly concerned that they operate differently. Specifically that `adjustFontSize` with `delta:5` is a relative change, whereas `adjustOpacity` with `opacity:5` is a "set" change. Maybe `relative` should default to `true`?

CC @cinnamon-msft for design related question

File: src/renderer/dx/DxRenderer.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/renderer/dx/DxRenderer.cpp:921]: God, that's a bad merge. Frick

**Comment by zadjii-msft** [comment]
> Specifically that `adjustFontSize` with `delta:5` is a relative change, whereas `adjustOpacity` with `opacity:5` is a "set" change. Maybe `relative` should default to `true`?

That's the one I'm most amenable to here. 



> can we update `adjustFontSize` to operate the same way?

Already planning on it - I want a bunch of actions that are good for previewing, so I'm thinking of adding a pile of `adjustFontSize(8)`, `adjustFontSize(12)`, `adjustFontSize(16)`, `adjustFontSize(24)` (etc) entries to a "Set Font Size..." menu in the command palette

**Comment by DHowett** [comment]
> Already planning on it - I want a bunch of actions that are good for previewing, so I'm thinking of adding a pile of `adjustFontSize(8)`, `adjustFontSize(12)`, `adjustFontSize(16)`, `adjustFontSize(24)` (etc) entries to a "Set Font Size..." menu in the command palette



hmmmmmm

when I see things like this, i feel like we're missing a layer of abstraction -- each one of these is an individual action, each one shows up in the list of bindable actions, and each one is a bit more cognitive overhead.

Are we doing that *just* to make it previewable?

**Comment by zadjii-msft** [comment]
discussion: we're defaulting to `relative:true`

**Review by carlos-zamora — APPROVED** [review]
Looks great! Want to file a follow-up for updating `adjustFontSize`? Ideally, that change would be released alongside this one imo.

File: src/cascadia/TerminalControl/TermControl.idl [hunk 8]
```diff
@@ -72,5 +72,14 @@ namespace Microsoft.Terminal.Control
         void ToggleReadOnly();
 
         String ReadEntireBuffer();
+
+        void AdjustOpacity(Int32 Opacity, Boolean relative);
+
+        // You'd think this should just be "Opacity", but UIElement already
```
Comment (miniksa) [inline src/cascadia/TerminalControl/TermControl.idl:78]: Good call on leaving this comment.

**Review by miniksa — COMMENTED** [review]
Yep, looks good per our discussion in the team meeting on Monday.

**Review by miniksa — APPROVED** [review]

**Comment by zadjii-msft** [comment]
> Want to file a follow-up for updating `adjustFontSize`? Ideally, that change would be released alongside this one imo.

#9818 notes this. I suppose I can work on that. Easy enough.

**Pull request merged**